File: flutter/android/app/src/main/kotlin/io/mindnav/mind_nav/MindNavDeviceHarnessPlugin.kt
package io.mindnav.mind_nav

import android.content.Context
import android.app.ActivityManager
import android.os.Build
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager
import android.util.Log
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class MindNavDeviceHarnessPlugin : FlutterPlugin, MethodChannel.MethodCallHandler {
    private lateinit var context: Context
    private lateinit var channel: MethodChannel

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, "mindnav.dev/device_harness")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        Log.d("MindNavDevice", "device bridge request: ${call.method}")
        when (call.method) {
            "capabilities" -> result.success(mapOf(
                "hapticsAvailable" to vibrator().hasVibrator(),
                "platform" to "android",
                "totalMemoryMiB" to totalMemoryMiB(),
            ))
            "acknowledgeTurn" -> {
                val vibrator = vibrator()
                if (vibrator.hasVibrator()) {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                        vibrator.vibrate(VibrationEffect.createOneShot(16, VibrationEffect.DEFAULT_AMPLITUDE))
                    } else {
                        @Suppress("DEPRECATION")
                        vibrator.vibrate(16)
                    }
                }
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    @Suppress("DEPRECATION")
    private fun vibrator(): Vibrator = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
        (context.getSystemService(Context.VIBRATOR_MANAGER_SERVICE) as VibratorManager).defaultVibrator
    } else {
        context.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
    }

    private fun totalMemoryMiB(): Int {
        val info = ActivityManager.MemoryInfo()
        (context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager).getMemoryInfo(info)
        return (info.totalMem / (1024L * 1024L)).toInt()
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
}

File: flutter/ios/Runner/MindNavDeviceHarnessPlugin.swift
import Flutter
import UIKit

final class MindNavDeviceHarnessPlugin: NSObject, FlutterPlugin {
  static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "mindnav.dev/device_harness",
      binaryMessenger: registrar.messenger()
    )
    registrar.addMethodCallDelegate(MindNavDeviceHarnessPlugin(), channel: channel)
  }

  func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "capabilities":
      result([
        "hapticsAvailable": true,
        "platform": "ios",
        "totalMemoryMiB": Int(ProcessInfo.processInfo.physicalMemory / (1024 * 1024)),
      ])
    case "acknowledgeTurn":
      let feedback = UIImpactFeedbackGenerator(style: .soft)
      feedback.prepare()
      feedback.impactOccurred(intensity: 0.35)
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}
